crates/rust-client/src/builder.rs

```
@@ -0,0 +1,187 @@
+    #[must_use]
+    pub fn in_debug_mode(mut self, debug: bool) -> Self {
+        self.in_debug_mode = debug;
+        self
+    }
+
+    /// Build and return the `Client`.
+    ///
+    /// # Errors
+    ///
+    /// - Returns an error if no RPC client or endpoint was provided.
+    /// - Returns an error if the store cannot be instantiated or the keystore fails.
+    pub async fn build(self) -> Result<Client<RpoRandomCoin>, ClientError> {
+        // Determine the RPC client to use.
+        let rpc_api: Box<dyn NodeRpcClient + Send> = if let Some(client) = self.rpc_api {
+            client
+        } else if let Some(endpoint) = self.rpc_endpoint {
+            Box::new(TonicRpcClient::new(&endpoint, self.timeout_ms))
+        } else {
+            return Err(ClientError::ClientInitializationError(
+                "RPC client or endpoint is required. Call `.with_rpc(...)` or `.with_rpc_client(...)`.".into(),
+            ));
+        };
+
+        // If no store was provided, create a SQLite store from the given path.
+        let arc_store: Arc<dyn Store> = if let Some(store) = self.store {
+            store
+        } else {
+            let store = SqliteStore::new(self.store_path.clone().into())
+                .await
+                .map_err(ClientError::StoreError)?;
+            Arc::new(store)
+        };
+
+        // Use the provided RNG, or create a default one.
+        let rng = if let Some(user_rng) = self.rng {
+            user_rng
+        } else {
+            let mut seed_rng = rand::thread_rng();
+            let coin_seed: [u64; 4] = seed_rng.gen();
+            RpoRandomCoin::new(coin_seed.map(Felt::new))
+        };
+
+        // Use the provided keystore, or default to a FilesystemKeyStore with temp_dir.
+        let keystore = match self.keystore {
+            Some(ks) => ks,
+            None => FilesystemKeyStore::new(temp_dir())
+                .map_err(|err| ClientError::ClientInitializationError(err.to_string()))?,
```

Should we fail if the keystore is not set? If not, I wouldn't use the `temp_dir` to store the keys, it might get cleared up.

---

Ah yes that is true. What would be a good reasonable default? I think I was following the cli tests as a guide when writing this, that's why I use `temp_dir()`.

---

In the CLI we create a "keystore" dir on the current directory and use that. 

But in this case, I feel like we should have it so that the user always provides the keystore themselves. My reasoning is that if we create the keystore in the builder, then the user won't have a reference to it. This means that when adding new accounts, the user won't be able to add the authentication keys to the keystore.

---

Would this make sense as a standard use example? This creates a `/keystore` directory.

```rust
let keystore = FilesystemKeyStore::new("keystore".into()).unwrap();
let mut client = ClientBuilder::new()
    .with_rpc(Endpoint::try_from("https://rpc.testnet.miden.io:443").unwrap())
    .with_timeout(10_000)
    .with_keystore(keystore)
    .with_sqlite_store("store.sqlite3")
    .in_debug_mode(true)
    .build()
    .await?;
```

---

This way the user has reference to the keystore.

---

If we make it so that the `with_keystore()` method takes a string, then we would have to change it so that the builder also returns the keystore it created internally, something like:
```
let (mut client, mut keystore) = ClientBuilder::new()
    .with_rpc(Endpoint::try_from("https://rpc.testnet.miden.io:443").unwrap())
    .with_timeout(10_000)
    .with_keystore("keystore")
    .with_sqlite_store("store.sqlite3")
    .in_debug_mode(true)
    .build()
    .await?;
```
I'm open to either one but I think I prefer the first one.

---

I think one of the goals here is to make assumptions about a reasonable set of defaults while also keeping some flexibility. So, I'd probably go with more specialized methods that make ergonomics a bit better. For example:
```rust
let mut client = ClientBuilder::new()
    .with_tonic_rpc("https://rpc.testnet.miden.io:443")
    .with_timeout(10_000)
    .with_filesystem_keystore("keystore")
    .with_sqlite_store("store.sqlite3")
    .in_debug_mode(true)
    .build()
    .await?;
```
Ff the user wanted to customize things beyond these defaults, they could also do something like:
```rust
let mut client = ClientBuilder::new()
    .with_rpc(my_rpc)
    .with_keystore(my_keystore)
    .with_store(my_store)
    .in_debug_mode(true)
    .build()
    .await?;
```
The only thing which I think doesn't fit this pattern would be the random coin. But maybe we keep it as is for now and make it more flexible in another PR.